### PR TITLE
[WIP] [Improve] optimize user locking, resource transfer, and user unlocking

### PR DESCRIPTION
**Pull request by VampireAchao**
1. Keep resource transfer and user lock as one transaction.
2. Optimize the front-end interaction for user locking/unlocking.
3. Any operation is disabled instantly after the user is locked.
4. The prompt message in the transfer dialog box is changed to be shorter.
5. Users cannot lock themselves.
6. The resources can only be transferred to users in the same team.

### CODE REVIEW

**Comment by wolfboys** [comment]
Thanks for your contribution, I will review it later

**Comment by zhoulii** [comment]
Hi @VampireAchao,  Thanks for your contribution, I build and deploy sp with this commit, and the fe cannot be visited. 

Snipaste_2023-05-11_14-00-40

This may be related to [UserStatusCheckInterceptor.java#L40](https://github.com/apache/incubator-streampark/blob/0f9b849a574b27676ad55d8ead96692a281b1d85/streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/base/interceptor/UserStatusCheckInterceptor.java#L40). I will continue the review when this issue is fixed.🙂

**Comment by VampireAchao** [comment]
> Hi @VampireAchao, Thanks for your contribution, I build and deploy sp with this commit, and the fe cannot be visited.
> 
> Snipaste_2023-05-11_14-00-40
> 
> This may be related to [UserStatusCheckInterceptor.java#L40](https://github.com/apache/incubator-streampark/blob/0f9b849a574b27676ad55d8ead96692a281b1d85/streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/base/interceptor/UserStatusCheckInterceptor.java#L40). I will continue the review when this issue is fixed.🙂

Thx for ur review, I'll fix ASAP.

**Comment by wolfboys** [comment]
Thanks for your hard work, I will review later

**Comment by zhoulii** [comment]
Hi @VampireAchao, Thanks for your contribution, the create user function is not available in this pr, can you fix it ? I will review later.

image

**Comment by VampireAchao** [comment]
> Hi @VampireAchao, Thanks for your contribution, the create user function is not available in this pr, can you fix it ? I will review later.
> 
> image

Hang on

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/base/exception/UserLogoutException.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.streampark.console.base.exception;
+
+public class UserLogoutException extends ApiAlertException {
```
Comment (zhoulii) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/base/exception/UserLogoutException.java]:
```suggestion
public class UserLockedException extends ApiAlertException {
```

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/system/service/impl/UserServiceImpl.java [hunk 2]
```diff
@@ -263,8 +251,31 @@ public Map<String, Object> generateFrontendUserInfo(User user, Long teamId, JWTT
 
   @Override
   @Transactional(rollbackFor = Exception.class)
-  public void transferResource(Long userId, Long targetUserId) {
-    applicationService.changeOwnership(userId, targetUserId);
-    resourceService.changeOwnership(userId, targetUserId);
+  public boolean lockUser(Long userId, Long transferToUserId) {
+    boolean hasResource =
+        applicationService.existsByUserId(userId) || resourceService.existsByUserId(userId);
```
Comment (zhoulii) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/system/service/impl/UserServiceImpl.java]: There are some other resources binding to a user, like variable, alertconfig...

File: streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/system/service/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -263,8 +251,31 @@ public Map<String, Object> generateFrontendUserInfo(User user, Long teamId, JWTT
...
+    if (hasResource && transferToUserId == null) {
+      return true;
+    }
+    if (transferToUserId != null) {
```
Comment (zhoulii) [inline streampark-console/streampark-console-service/src/main/java/org/apache/streampark/console/system/service/impl/UserServiceImpl.java:286]: `userId` and `transferToUserId` cannot be same,  and `transferToUserId` cannot be a locked user.

File: streampark-console/streampark-console-webapp/src/locales/lang/en/system/user.ts [hunk 4]
```diff
@@ -21,6 +21,11 @@ export default {
     modify: 'modify user',
     reset: 'reset password',
     resetTip: 'reset password, are you sure',
+    lock: 'lock user',
+    lockTip: 'lock user, are you sure',
```
Comment (zhoulii) [inline streampark-console/streampark-console-webapp/src/locales/lang/en/system/user.ts]:
```suggestion
    lockTip: 'Are you sure to lock this user ?',
```

File: streampark-console/streampark-console-webapp/src/locales/lang/en/system/user.ts [hunk 5]
```diff
@@ -44,7 +49,7 @@ export default {
     edit: 'Edit User',
     view: 'View User',
     notice: 'Notice',
-    transferResource: 'Please transfer the resources of the user who needs to be disabled to a new user'
+    transferResource: 'Please transfer the resources to a new user',
```
Comment (zhoulii) [inline streampark-console/streampark-console-webapp/src/locales/lang/en/system/user.ts]:
```suggestion
    transferResource: 'Please transfer the resources to another user',
```

File: streampark-console/streampark-console-webapp/src/locales/lang/zh-CN/system/user.ts [hunk 6]
```diff
@@ -44,7 +49,7 @@ export default {
     edit: '编辑用户',
     view: '查看用户',
     notice: '提示',
-    transferResource: '请将需要被禁用的用户资源转移到新的用户上'
+    transferResource: '请将资源转移到新的用户上',
```
Comment (zhoulii) [inline streampark-console/streampark-console-webapp/src/locales/lang/zh-CN/system/user.ts]:
```suggestion
    transferResource: '请将资源转移到其他用户',
```

File: streampark-console/streampark-console-webapp/src/views/system/user/User.vue [hunk 7]
```diff
@@ -82,11 +102,48 @@
...
+              label: t('system.user.form.transferResource'),
+              component: 'ApiSelect',
+              componentProps: {
+                api: async () => {
+                  let { records } = await fetchMemberList({
+                    page: 1,
+                    pageSize: 999999,
+                    teamId: userStore.getTeamId || '',
+                  });
+                  return records.filter((user) => user.userName !== userName.value);
```
Comment (zhoulii) [inline streampark-console/streampark-console-webapp/src/views/system/user/User.vue]: userName.value is the current logined user, not the user to be locked. Besides, we should also exclude users who are already locked.

File: streampark-console/streampark-console-webapp/src/views/system/user/User.vue [hunk 8]
```diff
@@ -113,6 +170,24 @@
               confirm: handleReset.bind(null, record),
             },
           },
+          {
+            icon: 'ant-design:lock-outlined',
+            color: 'error',
+            auth: 'user:delete',
+            ifShow: record.username !== 'admin' && record.status === '1',
```
Comment (zhoulii) [inline streampark-console/streampark-console-webapp/src/views/system/user/User.vue:178]: We'd better hide lock button for the logined user

**Review by zhoulii — COMMENTED** [review]
Thanks for your hard work, there are some minor problems that need to be addressed.

**Comment by VampireAchao** [comment]
Thanks for the review. I'll make the code changes later.

**Comment by wolfboys** [comment]
There are numerous conflicting files. I suggest initiating a new discussion and resubmitting the PR

**Pull request closed without merging**